File: trees_and_graphs/minimum_absolute_difference_in_bst.java
public List<Integer> dfs(TreeNode node)
{
    if (node == null)
    {
        return new ArrayList<>();
    }

    List<Integer> left = dfs(node.left);
    List<Integer> right = dfs(node.right);

    left.add(node.val);
    left.addAll(right);
    return left;
}

public int getMinimumDifferenceRecursive(TreeNode root)
{
    List<Integer> values = dfs(root);
    int ans = Integer.MAX_VALUE;
    for (int i = 1; i < values.size(); i++)
    {
        ans = Math.min(ans, values.get(i) - values.get(i - 1));
    }

    return ans;
}

public List<Integer> iterativeInorder(TreeNode root)
{
    Stack<TreeNode> stack = new Stack<>();
    List<Integer> values = new ArrayList<>();
    TreeNode curr = root;

    while (!stack.isEmpty() || curr != null)
    {
        if (curr != null)
        {
            stack.push(curr);
            curr = curr.left;
        }
        else
        {
            curr = stack.pop();
            values.add(curr.val);
            curr = curr.right;
        }
    }

    return values;
}

public int getMinimumDifferenceIterative(TreeNode root)
{
    List<Integer> values = iterativeInorder(root);
    int ans = Integer.MAX_VALUE;
    for (int i = 1; i < values.size(); i++)
    {
        ans = Math.min(ans, values.get(i) - values.get(i - 1));
    }

    return ans;
}

File: trees_and_graphs/minimum_absolute_difference_in_bst.cpp
#include "treenode.hpp"

#include <gtest/gtest.h>

#include <algorithm>
#include <limits>
#include <vector>

//! @brief Recursive helper function to get sorted vector of values in tree
//! @param[in] node Pointer to current node in tree
//! @return Vector of values from BST in sorted order
static std::vector<int> dfs(TreeNode* node)
{
    //! @details Time complexity O(N)

    if (node == nullptr)
    {
        return {};
    }

    auto       left  = dfs(node->left);
    const auto right = dfs(node->right);
    left.push_back(node->val);
    left.insert(left.end(), right.begin(), right.end());
    return left;
}

//! @brief Get min abs diff between values of any two nodes in BST
//! @param[in] root Pointer to root of BST
//! @return Minimum absolute difference between values of any two nodes in tree
static int getMinimumDifferenceRecursive(TreeNode* root)
{
    auto values = dfs(root);
    int  ans {std::numeric_limits<int>::max()};
    for (int i = 1; i < static_cast<int>(values.size()); ++i)
    {
        ans = std::min(ans, values[i] - values[i - 1]);
    }

    return ans;
}

TEST(GetMinimumDifferenceTest, SampleTest)
{
    TreeNode five {5};
    TreeNode nine {9};

    const TreeNode one {1};
    const TreeNode seven {7};
    const TreeNode fifteen {15};

    nine.left  = &five;
    nine.right = &fifteen;

    five.left  = &one;
    five.right = &seven;

    EXPECT_EQ(2, getMinimumDifferenceRecursive(&nine));
}